synapse/events/third_party_rules.py

```diff
@@ -46,6 +125,43 @@ def __init__(self, hs: "HomeServer"):
                 module_api=hs.get_module_api(),
             )
 
+        self._check_event_allowed_callbacks: List[CHECK_EVENT_ALLOWED_CALLBACK] = []
+        self._on_create_room_callbacks: List[ON_CREATE_ROOM_CALLBACK] = []
+        self._check_threepid_can_be_invited_callbacks: List[
+            CHECK_THREEPID_CAN_BE_INVITED_CALLBACK
+        ] = []
+        self._check_visibility_can_be_modified_callbacks: List[
+            CHECK_VISIBILITY_CAN_BE_MODIFIED_CALLBACK
+        ] = []
+
+    def register_third_party_rules_callbacks(
+        self,
+        check_event_allowed: Optional[CHECK_EVENT_ALLOWED_CALLBACK] = None,
+        on_create_room: Optional[ON_CREATE_ROOM_CALLBACK] = None,
+        check_threepid_can_be_invited: Optional[
+            CHECK_THREEPID_CAN_BE_INVITED_CALLBACK
+        ] = None,
+        check_visibility_can_be_modified: Optional[
+            CHECK_VISIBILITY_CAN_BE_MODIFIED_CALLBACK
+        ] = None,
```

Comment:
This list might grow very large at some point, and I'm aware that there's a few different places one needs to edit to add a new method to the interface.

I think this could all be replaced by a `Dict[str, List[Callable]]` that looks like:

```jsonc
{
    "check_event_allowed": [callback1, callback2]
    "on_create_room": [callback3, callback4],
    // ...
}
```

Not that this needs to be done now, but it might make things simpler for the future. I'm also not sure whether this setup would preserve the same type-checking we currently have though.

It would also allow for a simpler implementation of this method:

```python
def register_third_party_rules_callbacks(callback_names_to_functions):
    for name, f in callback_names_to_functions:
        if name not in self._callbacks:
            raise Exception("Unknown callback")

        self._callbacks[name].append(f)
```

Reply:
This reproducing the way spam checker callbacks are registered. Since we're trying to standardise the module interface, I'd like it if we can avoid having different ways to register callbacks depending on the module type. There's more context in https://github.com/matrix-org/synapse/pull/10062 on why we're doing it like that, but the gist of it is that it's the only way to have good enough typing of the callbacks without going for something over-engineered.